pilot/pkg/config/coredatamodel/discovery.go

```diff
@@ -81,132 +150,138 @@ func (d *MCPDiscovery) Services() ([]*model.Service, error) {
+	services := convertServices(cfg)
+	for _, service := range services {
+		for _, serviceEntryPort := range serviceEntry.Ports {
+			if len(serviceEntry.Endpoints) == 0 &&
+				serviceEntry.Resolution == networking.ServiceEntry_DNS {
+				// when service entry has discovery type DNS and no endpoints
+				// we create endpoints from service's host
+				// Do not use serviceentry.hosts as a service entry is converted into
+				// multiple services (one for each host)
+				// NOTE: these are excluded from byIP since GetProxyServiceInstances
+				// can not work on hostnames
+				svcInstance := &model.ServiceInstance{
+					Endpoint: model.NetworkEndpoint{
+						UID:         parseUID(cfg),
+						Address:     string(service.Hostname),
+						Port:        int(serviceEntryPort.Number),
+						ServicePort: convertPort(serviceEntryPort),
+					},
+					Service: service,
+					Labels:  nil,
+					TLSMode: model.DisabledTLSModeLabel,
+				}
+				if svcInstances, exist := byHost[service.Hostname]; exist {
+					svcInstances = append(svcInstances, svcInstance)
+					byHost[service.Hostname] = svcInstances
+				} else {
+					byHost[service.Hostname] = []*model.ServiceInstance{svcInstance}
+				}
+			} else {
+				for _, endpoint := range serviceEntry.Endpoints {
+					svcInstance := convertEndpoint(cfg, service, serviceEntryPort, endpoint)
+					// populate byIP with all endpoints attached to service
+					if svcInstances, exist := byIP[endpoint.Address]; exist {
+						svcInstances = append(svcInstances, svcInstance)
+						byIP[endpoint.Address] = svcInstances
+					} else {
+						byIP[endpoint.Address] = []*model.ServiceInstance{svcInstance}
+					}
+					// populate byHost with all service instances
+					if svcInstances, exist := byHost[service.Hostname]; exist {
+						svcInstances = append(svcInstances, svcInstance)
+						byHost[service.Hostname] = svcInstances
+					} else {
+						byHost[service.Hostname] = []*model.ServiceInstance{svcInstance}
+					}
+
+				}
+				notReadyEps := notReadyEndpoints(cfg)
+				for ip, port := range notReadyEps {
+					svcInstancesFromNotReadyEps := convertNotReadyEndpoints(service, serviceEntryPort, ip, port)
+					// populate byIP with notReadyEndpoints associated to service
+					if svcInstances, exist := byIP[ip]; exist {
+						svcInstances = append(svcInstances, svcInstancesFromNotReadyEps...)
+						byIP[ip] = svcInstances
+					} else {
+						byIP[ip] = svcInstancesFromNotReadyEps
+					}
+					// populate byHost with notReadyEndpoints associated to service
+					if svcInstances, exist := byHost[service.Hostname]; exist {
+						svcInstances = append(svcInstances, svcInstancesFromNotReadyEps...)
+						byHost[service.Hostname] = svcInstances
+					} else {
+						byHost[service.Hostname] = svcInstancesFromNotReadyEps
 					}
 				}
 			}
 		}
 	}
-	return out, nil
+	return byIP, byHost
+}
+
+func parseUID(cfg model.Config) string {
+	return "kubernetes://" + cfg.Name + "." + cfg.Namespace
```

Comment:
It seems only when mixer enabled , this is set. I am not sure why.

Reply:
I will put the condition back to check if mixer is enabled and then set the UID.